### ISSUE TITLE
[Explore]: Collect K8s metrics based on Metric Proxy

### ISSUE THREAD

**Issue by acosta11** [cloudfoundry/cf-k8s-controllers#459]
### Background

To provide users basic metrics, cf defines the log cache api and a configuration option to specify the log cache endpoint. Log cache is responsible for returning the logs in the cf envelope format, so our dependency on the backing implementation for this functionality has also motivated an explore for handling the conversion of kubectl logs to cf envelopes as a potential workaround to using log cache itself.

As a known working solution,  metric-proxy provides a deployment backed by metrics-server for aggregation of kubernetes metrics through the log cache api. See if this solution continues to slot in directly with cf-k8s-api and the cf cli. Because cloud controller implements some of the handling logic itself for metrics, we may be able to take the underlying metric-proxy logic and directly implement the metrics fetching in cf-k8s-api.

If direct implementation does not work, try deploying the metric-proxy component and see if we can consume the api in some capacity through a stubbed implementation of the stats endpoint (likely using https://github.com/cloudfoundry/go-log-cache as the client).

### Acceptance Criteria

If direct implementation works, a spike branch.

If the direct implementation does not work but deployment does, document the recommended configuration and installation steps for metric-proxy. We will follow up with a story to integrate the component more rigorously into the cf-k8s-api stats endpoint and installation scripts.

### Number of days this exploration is expected to take. This is what is “pointed”.

3

### Dev Notes

Links:
Log Cache API - https://github.com/cloudfoundry/log-cache-release/blob/develop/src/README.md

Metric Proxy - https://github.com/cloudfoundry/metric-proxy

CF for K8s reference deployment - https://github.com/cloudfoundry/cf-for-k8s/tree/develop/config/metrics 

Existing grpc go client for log-cache - https://github.com/cloudfoundry/go-log-cache

Options:
Harvest code from Metric Proxy and directly integrate into cf-k8s-controllers
Deploy Metric Proxy and harvest a grpc go client for log-cache

**Comment by julian-hj** [comment]
Kind deployments don't have metrics server installed OOTB. We were able to get it up and running with the [standard configuration](https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml), after applying [this little patch](https://github.com/kubernetes-sigs/kind/issues/398#issuecomment-478311167) to the command args for the metrics server.

**Comment by julian-hj** [comment]
**TL;DR** we believe that the direct implementation approach will work just fine.

@matt-royal and I paired on a spike to just glue the metrics server client code directly into the pod repository code in the API shim as a way of fetching more metrics.

So far, we created some very hacky code that gets the CPU and memory usage for each pod and spits it out to logs. (we didn't bother to plumb it out through the repository interface yet. That code is in [this branch](https://github.com/cloudfoundry/cf-k8s-controllers/tree/issues/459-wip-metrics-server-client).

**Some things to note**:
- Disk usage gets collected a little differently. The metrics proxy does that [here](https://github.com/cloudfoundry/metric-proxy/blob/main/pkg/metrics/diskusage/fetcher.go). We think that can just be cribbed in the same way the other code was cribbed.
- We are just using the service account in the pod. Probably we want to add the unprivileged user somehow?
- CF [redacts some stats for particular users](https://v3-apidocs.cloudfoundry.org/version/3.113.0/#get-stats-for-a-process). Unclear if we need to get that fancy. 
- CF returns [a lot of metrics](https://v3-apidocs.cloudfoundry.org/version/3.113.0/#the-process-stats-object). We aren't sure which of these is important.
- K9s returns something in its usage table that is oddly 10x more than the value we get from `kubectl top`. We saw that the metrics we get back are the same as the `kubectl top` ones. We don't know what the column in k9s represents 🤷🏻‍♂️

**Comment by matt-royal** [comment]
We determined that the `instance_ports` and the various quota fields are not populated from metrics-proxy data. Given that, we believe that we've validated that we can inline the logic for memory, cpu and disk usage without issue.
